### PR TITLE
add stub libvmemcache API and helper files from PMDK repo

### PR DESCRIPTION
**Pull request by ldorau**
add stub libvmemcache API and helper files from PMDK repo:
1) modified existing files:
	src/CMakeLists.txt
	src/tests/test.c
	utils/check_license/file-exceptions.sh
	utils/check_whitespace

2) added files with stub libvmemcache API:
	src/libvmemcache.c
	src/libvmemcache.h
	src/libvmemcache.map
	src/vmemcache.c
	src/vmemcache.h
	src/vmemcache_heap.c
	src/vmemcache_heap.h
	src/vmemcache_repl.c
	src/vmemcache_repl.h

3) added helper files copied from PMDK common module:
	src/common.h (renamed from src/pmemcommon.h)
	src/file.c
	src/file.h
	src/file_posix.c
	src/mmap.c
	src/mmap.h
	src/mmap_posix.c
	src/os.h
	src/os_posix.c
	src/os_thread.h
	src/os_thread_posix.c
	src/out.c
	src/out.h
	src/sys/queue.h
	src/sys_util.h
	src/util.c
	src/util.h
	src/util_posix.c
	src/valgrind_internal.h
	src/vec.h
	src/vecq.h

4) added helper files copied from PMDK libpmemobj library:
	src/ravl.c
	src/ravl.h
---

This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/4)

### CODE REVIEW

**Comment by ldorau** [comment]
This PR contains PR https://github.com/pmem/vmemcache/pull/3 yet.

File: src/libvmemcache.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * version of the libvmemcache API as provided by this header file.
+ * Applications can verify that the version available at run-time
+ * is compatible with the version used at compile-time by passing
+ * these defines to vmemcache_check_version().
+ */
+#define VMEMCACHE_MAJOR_VERSION 1
+#define VMEMCACHE_MINOR_VERSION 1
+
+#ifndef _WIN32
+const char *vmemcache_check_version(unsigned major_required,
```
Comment (marcinslusarz) [inline src/libvmemcache.h]: I question the need for this API.

File: src/libvmemcache.h [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+const wchar_t *vmemcache_check_versionW(unsigned major_required,
+	unsigned minor_required);
+#endif
+
+/*
+ * Passing NULL to vmemcache_set_funcs() tells libvmemcache to continue to use
+ * the default for that function.  The replacement functions must not make
+ * calls back into libvmemcache.
+ */
+void vmemcache_set_funcs(
```
Comment (marcinslusarz) [inline src/libvmemcache.h]: Please no.

File: src/vmemcache.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define VMEMCACHE_PREFIX "libvmemcache"
+#define VMEMCACHE_LEVEL_VAR "VMEMCACHE_LEVEL"
+#define VMEMCACHE_FILE_VAR "VMEMCACHE_FILE"
+
+/* attributes of the vmemcache memory pool format for the pool header */
+#define VMEMCACHE_HDR_SIG "VMCACHE"	/* must be 8 bytes including '\0' */
```
Comment (marcinslusarz) [inline src/vmemcache.h]: Do we really need a header? This is volatile solution, so we can even use O_TMPFILE for pool file and never expose it as user-visible file.

File: src/libvmemcache.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+};
+
+typedef void vmemcache_on_evict(VMEMcache *cache,
+	const char *key, size_t key_size, void *arg);
+typedef int vmemcache_on_miss(VMEMcache *cache,
+	const char *key, size_t key_size, void *arg);
+
+#ifndef _WIN32
+VMEMcache *vmemcache_new(const char *path, size_t max_size, size_t segment_size,
+		enum vmemcache_replacement_policy replacement_policy);
```
[the remark this answers is not in the document]
Comment (pbalcer) [reply src/libvmemcache.h]: yes.

File: src/libvmemcache.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		size_t segment_size,
+		enum vmemcache_replacement_policy replacement_policy);
+VMEMcache *vmemcache_newW(const wchar_t *path, size_t max_size,
+		size_t segment_size,
+		enum vmemcache_replacement_policy replacement_policy);
+#endif
+
+void vmemcache_delete(VMEMcache *c);
+
+void vmemcache_callback_on_evict(VMEMcache *cache,
```
[the remark this answers is not in the document]
Comment (pbalcer) [reply src/libvmemcache.h]: I prefer this name.

File: src/libvmemcache.h (hunk 1 above)
Comment (pbalcer) [reply src/libvmemcache.h]: :+1:

File: src/libvmemcache.h (hunk 2 above)
Comment (pbalcer) [reply src/libvmemcache.h]: Yea, I guess we don't need it.

File: src/vmemcache.h (hunk 3 above)
Comment (pbalcer) [reply src/vmemcache.h]: :+1:

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -37,11 +37,26 @@ include(FindThreads)
...
+	mmap.c
+	mmap_posix.c
+	libvmemcache.c
+	ravl.c
+	vmemcache.c
+	vmemcache_heap.c
+	vmemcache_repl.c)
 
 add_library(vmemcache SHARED ${SOURCES})
-target_link_libraries(vmemcache PRIVATE -Wl,--version-script=${CMAKE_SOURCE_DIR}/src/libvmemcache.map)
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/CMakeLists.txt]: By accident, added.

File: Findvalgrind.cmake [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#
```
[the remark this answers is not in the document]
Comment (ldorau) [reply Findvalgrind.cmake]: Removed

File: src/libvmemcache.h [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	vmemcache_on_evict *evict, void *arg);
+void vmemcache_callback_on_miss(VMEMcache *cache,
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/libvmemcache.h]: .

File: src/libvmemcache.h [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+#define vmemcache_check_version vmemcache_check_versionU
+#define vmemcache_errormsg vmemcache_errormsgU
+
+#endif
+
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/libvmemcache.h]: Done

File: src/vmemcache.h (hunk 3 above)
Comment (ldorau) [reply src/vmemcache.h]: Removed

File: src/libvmemcache.h (hunk 2 above)
Comment (ldorau) [reply src/libvmemcache.h]: Removed

File: src/libvmemcache.h (hunk 1 above)
Comment (ldorau) [reply src/libvmemcache.h]: Removed

**Pull request merged**